### PR TITLE
PR-1852 - all state models should be configurable

### PR DESCRIPTION
**Pull request by EddiTim**

### CODE REVIEW

File: service/grails-app/services/org/olf/rs/statemodel/StatusService.groovy [hunk 1]
```diff
@@ -506,28 +506,29 @@ public class StatusService {
...
-            } else {
-                settingsKey = SettingsData.SETTING_STATE_MODEL_REQUESTER;
+                settingsKey = isRequester
+                        ? SettingsData.SETTING_STATE_MODEL_REQUESTER_NON_RETURNABLE
+                        : SettingsData.SETTING_STATE_MODEL_RESPONDER_NON_RETURNABLE
+            } else if (request.serviceType?.value == 'loan') {
+                settingsKey = isRequester
+                        ? SettingsData.SETTING_STATE_MODEL_REQUESTER_RETURNABLE
+                        : SettingsData.SETTING_STATE_MODEL_RESPONDER_RETURNABLE
+            } else if (request.deliveryMethod?.value == 'url') {
```
Comment (jakub-id) [inline service/grails-app/services/org/olf/rs/statemodel/StatusService.groovy:522]: This doesn't match the old logic and this branch will never execute. This condition should be under the`loan` condition. I would prefer you keeped the old logic as is:
```
if serviceType == copy {
    non-returnables
} else {
    if deliveryMethod == URL {
      digital-returnables
    } else {
       returnables
    }
}
```

File: service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy [hunk 2]
```diff
@@ -142,10 +142,16 @@ public class SettingsData {
     public static final String SETTING_NETWORK_TIMEOUT_PERIOD       = 'network_timeout_period';
 
     // State model configuration settings
-    public static final String SETTING_STATE_MODEL_REQUESTER     = 'state_model_requester';
-    public static final String SETTING_STATE_MODEL_REQUESTER_CDL = 'state_model_requester_cdl';
```
Comment (jakub-id) [inline service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy:146]: Was this static used for anything?

File: service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy [hunk 3]
```diff
@@ -142,10 +142,16 @@ public class SettingsData {
...
-    public static final String SETTING_STATE_MODEL_RESPONDER     = 'state_model_responder';
-    public static final String SETTING_STATE_MODEL_RESPONDER_CDL = 'state_model_responder_cdl';
```
Comment (jakub-id) [inline service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy:148]: Was this static used for anything?

**Review by jakub-id — CHANGES_REQUESTED** [review]

File: service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy (hunk 2 above)
Comment (EddiTim) [reply service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy:146]: No, I will remove it.

File: service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy (hunk 3 above)
Comment (EddiTim) [reply service/src/main/groovy/org/olf/rs/referenceData/SettingsData.groovy:148]: No, I will remove it.

**Comment by skomorokh** [comment]
It seems like this change prevents `StatusService.getStateModel` from ever resolving to the CDL responder model (and has no setting for which model to use for CDL responder)

**Comment by EddiTim** [comment]
> It seems like this change prevents `StatusService.getStateModel` from ever resolving to the CDL responder model (and has no setting for which model to use for CDL responder)

if (request.deliveryMethod?.value == 'url') {
                    settingsKey = isRequester
                            ? SettingsData.SETTING_STATE_MODEL_REQUESTER_DIGITAL_RETURNABLE
                            : SettingsData.SETTING_STATE_MODEL_RESPONDER_DIGITAL_RETURNABLE
                            
SETTING_STATE_MODEL_RESPONDER_DIGITAL_RETURNABLE = CDLResponder

**Comment by EddiTim** [comment]
> There is no single responder model for digital returnable. There is a CDL responder and there will be an eBook responder. They're both digital returnables but they have a different workflow on the supplier side. It's entirely possible the same tenant will use both so it's not just that they can set it to one or the other.
> 
> Whereas on the requester side, at least as initially envisioned, they have the same process and we group them together as digital returnable requester.

@jakub-id Please have a look and we need to decide what to do, as Jason is pointing out a valid thing

**Comment by jakub-id** [comment]
> There is no single responder model for digital returnable. There is a CDL responder and there will be an eBook responder. They're both digital returnables but they have a different workflow on the supplier side. It's entirely possible the same tenant will use both so it's not just that they can set it to one or the other.
> 
> Whereas on the requester side, at least as initially envisioned, they have the same process and we group them together as digital returnable requester.

I don't see that the old code supported two different digital returnables state models for the responder as it hard-coded the  CDL state model. I think we should go ahead and approve it as it allows configuring state models for ZFL and create a separate ticket for more granular configurability of the digital returnable state models.

**Comment by skomorokh** [comment]
Please use the name `SETTING_STATE_MODEL_RESPONDER_CDL` instead of `SETTING_STATE_MODEL_RESPONDER_DIGITAL_RETURNABLE` so that we can avoid having to rename it when we implement ebooks.

**Pull request merged**